### PR TITLE
Bug fixes for ESMF on macOS with apple-clang+gfortran, with gfortran-10+ in general, and on Cray

### LINKED ISSUE #29192: ESMF doesn't built on macOS with apple-clang+gfortran, doesn't built with gfortran-10+, and doesn't build on Cray
### Steps to reproduce

Installing ESMF fails on the following systems:
1. macOS with apple-clang+gfortran
2. Any system using gfortran-10+
3. Cray systems such as ORNL's Gaea system

To repeat:
```
spack install esmf
```
on any of these systems.

All of these problems are known and have been solved elsewhere (https://github.com/noaa-emc/hpc-stack, https://github.com/JCSDA/jedi-stack).

### Error message

Various error messages that are either related to gfortran-10+ (strict argument checking), the use of apple-clang instead of gcc or (llvm-)clang, or being on a cray. For cray, the error is in the final linker step and has to do with the MPI libraries.

I am not submitting the error logs here, because the problems and the solutions are well known and have been solved elsewhere (https://github.com/noaa-emc/hpc-stack, https://github.com/JCSDA/jedi-stack).

### Information on your system

1. macOS
```
* **Spack:** 0.17.1-1141-bda4216b9a
* **Python:** 3.9.10
* **Platform:** darwin-monterey-x86_64
* **Concretizer:** clingo
```
2. Cray
```
* **Spack:** 0.17.1-1141-bda4216b9a
* **Python:** 3.7.3
* **Platform:** cray-sles15-haswell
* **Concretizer:** clingo
```

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

### PR DESCRIPTION
**Pull request by climbfuji**
Fixes https://github.com/spack/spack/issues/29192 (ESMF doesn't built on macOS with apple-clang+gfortran, doesn't built with gfortran-10+, and doesn't build on Cray). These are the three changes related to the three errors in the issue:

1. Add the missing logic for `apple-clang`
2. Add the usual `-fallow-argument-mismatch`
3. For Cray, one needs to export `ESMF_OS=Unicos`

This has been tested on macOS with (llvm-)clang@13.0.0 + gfortran@11.2.0 and with apple-clang@13.0.0 + gfortran@11.2.0, as well as on ORNL's Gaea system.

### CODE REVIEW

**Comment by climbfuji** [comment]
@adamjstewart I'd be willing to be one of the maintainers of the ESMF package. Would you be interested in (a) reviewing this PR and/or (b) being added to the list of maintainers as part of this PR (or be the sole maintainer)?

Also, in https://github.com/NOAA-EMC/hpc-stack/blob/59e8a753227a86a0cdec9d16942dda6f71815eb4/libs/build_esmf.sh#L121 there is room for overwriting the `ESMF_OS=Unicos` setting. I think this is for some newer crays that identify themselves more like a linux system than a cray. Maybe we need to be able to distinguish between "cray crays" and "other crays".

**Comment by climbfuji** [comment]
@kgerheiser FYI.

File: var/spack/repos/builtin/packages/esmf/package.py [hunk 1]
```diff
@@ -126,10 +126,12 @@ def edit(self, spec, prefix):
         # C++ compilers are being used to build the ESMF library.
         if self.compiler.name == 'gcc':
             os.environ['ESMF_COMPILER'] = 'gfortran'
+            gfortran_major_version = int(spack.compiler.get_compiler_version_output(self.compiler.fc, '-dumpversion').split('.')[0])
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/esmf/package.py]: I feel like there has to be an easier way to get the compiler version than this. Does `self.compiler.version.up_to(1)` work?

**Review by adamjstewart — COMMENTED** [review]
Looks good to me other than the compiler version thing. I'm probably not the best person to be a maintainer for this package since I've never used it before, but feel free to add yourself!

File: var/spack/repos/builtin/packages/esmf/package.py (hunk 1 above)
Comment (climbfuji) [reply var/spack/repos/builtin/packages/esmf/package.py]: Thanks for taking the time to review the PR.

I can look into this, I had asked a question in the discussions but didn't get any response: https://github.com/spack/spack/discussions/28146

Ok, I can take care of ESMF for the moment, and add other maintainers later as required.

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/esmf/package.py]: Ah, yeah, that might be necessary for mixed compiler environments like Clang + GFortran. Let's leave that as is for now. We're planning a big overhaul that will make compilers formal dependencies and that should make it easier to express and query this information.

Thanks for volunteering to be a maintainer!

**Comment by climbfuji** [comment]
@adamjstewart I added the `maintainers` line as instructed by the bot.

**Review by adamjstewart — DISMISSED** [review]

**Review by adamjstewart — APPROVED** [review]

**Comment by climbfuji** [comment]
@adamjstewart this is ready to merge from my side, I am having a hard time catching up with updating from develop so either someone with superpowers need to merge or I'll keep trying.

**Comment by climbfuji** [comment]
Thanks, @adamjstewart !

**Pull request merged**